File: generate/src/log.js
var { Record, List, Map } = require("immutable");
var chance = require("chance").Chance();

var Datasets = require("./datasets");

var LogRecord = Record({
  nodeIp: "",
  timestamp: "",
  provider: "",
  clientIp: "",
  originServer: "",
  sessionId: "",
  channel: "",
  qualityId: "",
  fragmentId: "",
  userAgent: "",
  totalBytesSent: "",
  bytesSentFromCache: ""
});

class Log extends LogRecord {
  static create(timestamp) {
    var totalBytesSent = chance.integer({
      min: 1024,
      max: 2097152
    });
    var bytesSentFromCache = chance.integer({
      min: 0,
      max: totalBytesSent
    });
    var provider = chance.pick(Map(Datasets.provider).valueSeq().toJS());
    var channel = chance.pick(Map(Datasets.channel).get(provider));
    return new Log({
      nodeIp: chance.pick(Datasets.nodeIp),
      timestamp: timestamp,
      provider: provider,
      clientIp: chance.ip(),
      originServer: chance.pick(Datasets.originServer),
      sessionId: chance.hash(),
      channel: channel,
      qualityId: chance.integer({ min: 0, max: 20 }),
      fragmentId: chance.integer({ min: 0, max: 500 }),
      userAgent: chance.pick(Datasets.userAgent),
      totalBytesSent: totalBytesSent,
      bytesSentFromCache: bytesSentFromCache
    });
  }

  get url() {
    return List([
      this.originServer,
      this.sessionId,
      this.channel,
      this.qualityId,
      this.fragmentId
    ]).join("/");
  }

  toString() {
    return List([
      this.nodeIp,
      this.timestamp,
      this.provider,
      this.clientIp,
      this.url,
      this.userAgent,
      this.totalBytesSent,
      this.bytesSentFromCache
    ]).join("\t");
  }
}

module.exports = Log;

File: calculate/model/log.py
class Log:
    def __init__(self, line):
        data = line.split("\t")
        self.node_ip = data[0]
        self.timestamp = int(data[1])
        self.provider = data[2]
        self.client_ip = data[3]
        url = data[4]
        self.user_agent = data[5]
        self.total_bytes = int(data[6])
        self.cache_bytes = int(data[7])
        url_data = url.split("/")
        self.origin_server = url_data[0]
        self.session_id = url_data[1]
        self.channel = url_data[2]
        self.quality_id = int(url_data[3])
        self.fragment_id = int(url_data[4])
